### ISSUE TITLE
GenericRecord not iterable

### ISSUE THREAD

**Issue by mhowlett** [confluentinc/avro#8]
The `GenericRecord` class allows access to values by key, but it's not iterable, so you need to know the keys in advance. The Java implementation is the same.

But this seems like something you'd often want! Is there something I'm missing?

**Comment by blachniet** [comment]
**tldr;** You can generally get around this limitation by iterating over the schema's field names and acquiring values from the generic record using those names:

```csharp
var schema = (RecordSchema)Schema.Parse("...");
var record = new GenericRecord(schema);

foreach (var field in schema.Fields)
{
    if (record.TryGetValue(field.Name, out object value))
    {
        // Do something with the value
    }
}
```

---

I've wanted this in a few scenarios as well. Since `GenericRecord` uses an `Dictionary<string, object>` (`contents`) under the hood, it *should* be easy to implement. We can simply return that dictionary's enumerator.

1. Indicate that we implement `IEnumerable<KeyValuePair<string, object>>` and `IEnumerable`

```csharp
    public class GenericRecord : IEnumerable<KeyValuePair<string, object>>, IEnumerable
    {
```

2. Implement the two methods

```csharp
        public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
        {
            return contents.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return contents.GetEnumerator();
        }
```

**However**, after trying to implement this myself, I noticed some unexpected behavior. When you create a new `GenericRecord`, it doesn't actually add anything to the `contents` dictionary until you call `Add`. This means that you won't see any of the fields from the schema until you've actually set/added them.

```json
{
  "type": "record", 
  "name": "myRecord", 
  "fields": [{
    "name": "f1",
    "type": "string"
  }]
}
```

```csharp
var record = new GenericRecord(myRecordSchema);
// record.contents is empty (record.ToList().Length == 0)

record.Add("f1", "my field value");
// record.contents has one value (record.ToList().Length == 1)
```

IMHO, this seems a little unintuitive. Maybe that's why it was left out initially. Maybe I'm taking too simplistic of an approach to solve the problem though. I guess we could iterate over the schema's field names inside the `GetEnumerator` method (just thought of that while typing up this post). I may think on this a bit more.

The approach used in the **tldr;** section has covered my use cases so far.
